### PR TITLE
Add the slugify filter 

### PR DESCRIPTION
**Pull request by jonathan-s**
Supersedes #61 as it was easier to do this in new commits rather than rebase the thing. 

Also an open question regarding file structure. When working through the files I just noticed that we have `render/filters.rs` and `filters.rs`, to me it would make sense to keep this in the same place. Or did you have any particular reasons for splitting that up?

### CODE REVIEW

**Comment by LilyFoote** [comment]
> Also an open question regarding file structure. When working through the files I just noticed that we have `render/filters.rs` and `filters.rs`, to me it would make sense to keep this in the same place. Or did you have any particular reasons for splitting that up?

I like having a split between code relevant to parsing and code relevant to rendering. Probably `filters.rs` should be `parse/filters.rs`, but I haven't dug into refactoring `parse.rs` yet.

**Comment by jonathan-s** [comment]
Thanks for the context!

File: src/render/filters.rs [hunk 1]
```diff
@@ -258,6 +270,35 @@ impl ResolveFilter for SafeFilter {
...
+                let content = NON_WORD_RE.replace_all(&content, "");
+                let content = content.trim();
+                let content = WHITESPACE_RE.replace_all(&content, "-");
+                content.to_string().into_content()
+            }
+            None => "".as_content(),
+        };
+        Ok(content)
+    }
+}
```
Comment (LilyFoote) [inline src/render/filters.rs:300]: Two suggestions in one here:

1. Pull out a `slugify` helper function.
2. Handle the different `content` variants explicitly, so we can avoid converting into Python unnecessarily.

I also discovered the existence of [`GILOnceCell::import`](https://docs.rs/pyo3/latest/pyo3/sync/struct.GILOnceCell.html#method.import), which can be used to avoid importing `SafeData` more than once.

We'll need tests to cover the extra `Content` match arms.

```suggestion
fn slugify(content: Cow<str>) -> Cow<str> {
    let content = content
        .nfkd()
        // first decomposing characters, then only keeping
        // the ascii ones, filtering out diacritics for example.
        .filter(|c| c.is_ascii())
        .collect::<String>()
        .to_lowercase();
    let content = NON_WORD_RE.replace_all(&content, "");
    let content = content.trim();
    let content = WHITESPACE_RE.replace_all(&content, "-");
    Cow::Owned(content.to_string())
}

static SAFEDATA: GILOnceCell<Py<PyType>> = GILOnceCell::new();

impl ResolveFilter for SlugifyFilter {
    fn resolve<'t, 'py>(
        &self,
        variable: Option<Content<'t, 'py>>,
        py: Python<'py>,
        _template: TemplateString<'t>,
        _context: &mut Context,
    ) -> ResolveResult<'t, 'py> {
        let content = match variable {
            Some(content) => match content {
                Content::Py(content) => {
                    let slug = slugify(Cow::Owned(content.str()?.extract::<String>()?));
                    #[allow(non_snake_case)]
                    let SafeData = SAFEDATA.import(py, "django.utils.safestring", "SafeData")?;
                    match content.is_instance(SafeData)? {
                        true => Some(Content::HtmlSafe(slug)),
                        false => Some(Content::String(slug)),
                    }
                }
                Content::Int(content) => {
                    Some(Content::String(slugify(Cow::Owned(content.to_string()))))
                }
                Content::Float(content) => {
                    Some(Content::String(slugify(Cow::Owned(content.to_string()))))
                }
                Content::String(content) => Some(Content::String(slugify(content))),
                Content::HtmlSafe(content) => Some(Content::HtmlSafe(slugify(content))),
            },
            None => "".as_content(),
        };
        Ok(content)
    }
}
```

**Review by LilyFoote — COMMENTED** [review]
Thanks for persevering with this!

I think we can do better with avoiding unnecessary conversions to Python objects though:

File: src/render/filters.rs (hunk 1 above)
Comment (jonathan-s) [reply src/render/filters.rs]: Thanks, I've applied this suggestion, and will make some refinements + add tests for the content match arms.

File: src/render/filters.rs [hunk 2]
```diff
@@ -279,20 +297,22 @@ impl ResolveFilter for SlugifyFilter {
...
+                Content::Py(content) => {
+                    let slug = slugify(Cow::Owned(content.str()?.extract::<String>()?));
+                    #[allow(non_snake_case)]
+                    let SafeData = SAFEDATA.import(py, "django.utils.safestring", "SafeData")?;
+                    match content.is_instance(SafeData)? {
+                        true => Some(Content::HtmlSafe(slug)),
+                        false => Some(Content::String(slug)),
+                    }
+                }
+                // Int and Float requires no slugify, we only need to turn it into a string.
```
Comment (LilyFoote) [inline src/render/filters.rs:310]: Good point!

File: src/render/filters.rs [hunk 3]
```diff
@@ -401,6 +401,15 @@ mod tests {
 
             assert_eq!(result, "a-b");
 
+            let engine = EngineData::empty();
+            let template_string = "{{ var|slugify }}".to_string();
+            let context = PyDict::new(py);
+            context.set_item("var", 1).unwrap();
```
Comment (jonathan-s) [inline src/render/filters.rs:407]: @LilyFoote It appears that even if we add an integer here this is still effectively treated as if this coming from the python arm. Do you have any thoughts around this?

Comment (LilyFoote) [reply src/render/filters.rs:407]: Yeah, `context` is basically a `HashMap<String, Py<PyAny>>`, so putting an integer in will convert to a Python `int`.

Instead, the way to do this is to chain multiple filters. I usually use `default` for this:

```
{{ var|default:1|slugify }}
```

The result of the `default` filter will be a `Content::Int(1)` if `var` is not defined.

Comment (jonathan-s) [reply src/render/filters.rs:407]: I see, that means that for the majority of cases this won't matter much. Though I guess the type of `HashMap` may change at some point to make this more relevant.

File: src/render/filters.rs [hunk 4]
```diff
@@ -299,6 +360,72 @@ mod tests {
...
+            context.set_item("var", "a & b").unwrap();
+            let template = Template::new_from_string(py, template_string, &engine).unwrap();
+            let result = template.render(py, Some(context), None).unwrap();
+
+            assert_eq!(result, "a-b");
+
+            let engine = EngineData::empty();
+            let template_string = "{{ var|default:1|slugify }}".to_string();
+            let context = PyDict::new(py);
+            context.set_item("var", 1).unwrap();
```
Comment (LilyFoote) [inline src/render/filters.rs:407]:
```suggestion
```

Comment (jonathan-s) [reply src/render/filters.rs]: Noticed :)

File: src/render/filters.rs [hunk 5]
```diff
@@ -299,6 +360,71 @@ mod tests {
         })
     }
 
+    #[test]
+    fn test_render_filter_slugify() {
```
Comment (LilyFoote) [inline src/render/filters.rs:364]: I think we should probably split this particular test up into multiple tests. If anything breaks for whatever reason, I'd rather see which of these scenarios fail rather than just one.

File: src/render/filters.rs [hunk 6]
```diff
@@ -299,6 +360,71 @@ mod tests {
...
+            assert_eq!(result, "hello-world");
+
+            let engine = EngineData::empty();
+            let template_string = "{{ var|slugify }}".to_string();
+            let context = PyDict::new(py);
+            context.set_item("var", " hello world").unwrap();
+            let template = Template::new_from_string(py, template_string, &engine).unwrap();
+            let result = template.render(py, Some(context), None).unwrap();
+
+            assert_eq!(result, "hello-world");
```
Comment (LilyFoote) [inline src/render/filters.rs:384]: Splitting the tests up would also allow the test name to point to the difference between these two tests - the leading whitespace in the second.

Comment (jonathan-s) [reply src/render/filters.rs]: Agreed, collecting them in the same test was partially laziness, partially keeping them a bit more organized compared to have single test functions. 

All the tests have been broken up now though.

File: src/render/filters.rs [hunk 7]
```diff
@@ -267,7 +328,23 @@ mod tests {
...
+            Some(mark_safe) => mark_safe,
+            None => {
+                let py_mark_safe = py.import("django.utils.safestring")?;
+                let py_mark_safe = py_mark_safe.getattr("mark_safe")?;
+                MARK_SAFE.set(py, py_mark_safe.into());
+                MARK_SAFE.get(py).unwrap()
+            }
+        };
+        let safe_string = mark_safe.call1(py, (string,))?;
+        Ok(safe_string)
```
Comment (jonathan-s) [inline src/render/filters.rs:346]: So generally, this would the pattern to call python functions from rust?

Comment (LilyFoote) [reply src/render/filters.rs:346]: I think in tests its a bit overkill to use `GILOnceCell` - that's more of an optimisation.

Comment (jonathan-s) [reply src/render/filters.rs:346]: Fair enough :). I'll leave it as is, as all things considering it works. 

Either way, this should be good to go now!

**Comment by jonathan-s** [comment]
🎉
